File: JSApp/SPA_ApplicationsAndApplicationsArchitecture/scripts/requester.js
var app = app || {};

app.requester = (function() {
    function AjaxRequester() {
        this.get = makeGetRequest;
        this.put = makePutRequest;
        this.post = makePostRequest;
        this.del = makeDeleteRequest;
    }

    var makeRequest = function(url, method, data, headers) {
        var def = Q.defer();

        $.ajax({
            headers: headers,
            url: url,
            method: method,
            contentType: 'application/json',
            data: JSON.stringify(data),
            success: function(data) {
                def.resolve(data);
            },
            error: function(error) {
                def.reject(error)
            }
        });

        return def.promise;
    }

    var makeHeaders = function(headers) {
        var personalHeader = JSON.parse(JSON.stringify(headers));
        personalHeader['X-Parse-Session-Token'] = localStorage.getItem('sessionToken');
        return personalHeader;
    }

    var makeGetRequest = function(url, headers) {
        return makeRequest(url, 'GET', null, makeHeaders(headers));
    }

    var makePostRequest = function(url, data, headers) {
        return makeRequest(url, 'POST', data, makeHeaders(headers));
    }

    var makePutRequest = function(url, data, headers) {
        return makeRequest(url, 'PUT', data, makeHeaders(headers));
    }

    var makeDeleteRequest = function(url, headers) {
        return makeRequest(url, 'DELETE', null, makeHeaders(headers));
    }

    return {
        get: function() {
            return new AjaxRequester();
        }
    }
})();

File: WebServicesAndCloud/2.AspWebApi/ConsumeServicesWithCSharp/Requester.cs
namespace ConsumeServicesWithCSharp
{
    using System;
    using System.Net.Http;
    using System.Net.Http.Headers;
    using System.Threading.Tasks;

    using Newtonsoft.Json;

    public static class Requester
    {
        private static HttpClient client; 

        static Requester()
        {
            client = new HttpClient();
            client.BaseAddress = new Uri(Utility.BaseUrl);
        }

        public static Task<string> MakeGetRequest(string serviceUrl)
        {
            var request = MakeRequest(serviceUrl, HttpMethod.Get);
            return client.SendAsync(request).ContinueWith(t =>
            {
                var response = t.Result;
                var content = response.Content.ReadAsStringAsync().Result;
                
                return content;
            });
        }

        private static HttpRequestMessage MakeRequest(string serviceUrl, HttpMethod method)
        {
            var request = new HttpRequestMessage
            {
                RequestUri = new Uri(Utility.BaseUrl + serviceUrl),
                Method = method,
            };
            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue(Utility.MediaType));
            return request;
        }

        public static Task<string> MakePostRequest(string url, object data)
        {
            var request = MakeRequest(url, HttpMethod.Post);
            var json = JsonConvert.SerializeObject(data);
            request.Content = new StringContent(json);
            request.Content.Headers.ContentType = new MediaTypeWithQualityHeaderValue(Utility.MediaType);

            return client.SendAsync(request).ContinueWith(
                t =>
                {
                    var response = t.Result;
                    return response.Content.ReadAsStringAsync().Result;
                });

        }

        public static Task<string> MakePutRequest(string url, object data)
        {
            var request = MakeRequest(url, HttpMethod.Put);
            var json = JsonConvert.SerializeObject(data);
            request.Content = new StringContent(json);
            request.Content.Headers.ContentType = new MediaTypeWithQualityHeaderValue(Utility.MediaType);
            return client.SendAsync(request).ContinueWith(
                t =>
                {
                    var response = t.Result;
                    return response.Content.ReadAsStringAsync().Result;
                });
        }

        public static Task<string> MakeDeleteRequest(string url)
        {
            var request = MakeRequest(url, HttpMethod.Delete);

            return client.SendAsync(request).ContinueWith(
                t =>
                {
                    var response = t.Result;
                    return response.Content.ReadAsStringAsync().Result;
                });
        }

    }
}
